diff --git a/src/compiler/AdaptiveCppLlvmPasses.cpp b/src/compiler/AdaptiveCppLlvmPasses.cpp
index 2af72ad..3bb0363 100644
--- a/src/compiler/AdaptiveCppLlvmPasses.cpp
+++ b/src/compiler/AdaptiveCppLlvmPasses.cpp
@@ -39,7 +39,11 @@
 #include "llvm/Pass.h"
 #include "llvm/Passes/PassBuilder.h"
+#if LLVM_VERSION_MAJOR >= 22
+#include "llvm/Plugins/PassPlugin.h"
+#else
 #include "llvm/Passes/PassPlugin.h"
+#endif
 #include "llvm/Support/CommandLine.h"

 #if LLVM_VERSION_MAJOR < 16
diff --git a/src/compiler/llvm-to-backend/metal/Emitter.cpp b/src/compiler/llvm-to-backend/metal/Emitter.cpp
index 7117f11..1ff61f0 100644
--- a/src/compiler/llvm-to-backend/metal/Emitter.cpp
+++ b/src/compiler/llvm-to-backend/metal/Emitter.cpp
@@ -31,7 +31,11 @@
 #include "llvm/IR/LegacyPassManager.h"
 #endif
 #include "llvm/MC/TargetRegistry.h"
+#if LLVM_VERSION_MAJOR >= 22
+#include "llvm/Plugins/PassPlugin.h"
+#else
 #include "llvm/Passes/PassPlugin.h"
+#endif
 #include "llvm/Support/Debug.h"
 #include "llvm/Support/ErrorHandling.h"
 #include "llvm/Support/FileSystem.h"
